cpu/sam0_common/periph/gpio.c

```diff
@@ -217,15 +227,51 @@ int gpio_init_int(gpio_t pin, gpio_mode_t mode, gpio_flank_t flank,
     _EIC->WAKEUP.reg |= (1 << exti);
     /* enable the EIC module*/
     _EIC->CTRL.reg = EIC_CTRL_ENABLE;
-    while (_EIC->STATUS.reg & EIC_STATUS_SYNCBUSY) {}
+    EIC_SYNC();
 #else /* CPU_FAM_SAML21 */
     /* enable the EIC module*/
     _EIC->CTRLA.reg = EIC_CTRLA_ENABLE;
-    while (_EIC->SYNCBUSY.reg & EIC_SYNCBUSY_ENABLE) {}
+    EIC_SYNC();
 #endif
     return 0;
 }
 
+inline static void reenable_eic(gpio_irq_clock_t clock) {
+#if defined(CPU_SAML21)
+    uint32_t ctrla_reg = EIC_CTRLA_ENABLE;
+
+    EIC->CTRLA.reg = 0;
+    EIC_SYNC();
+
+    if (clock == IRQ_CLOCK_SLOW) {
+        ctrla_reg |= EIC_CTRLA_CKSEL;
+    }
+
+    EIC->CTRLA.reg = ctrla_reg;
+    EIC_SYNC();
+#endif
+}
+
+void gpio_pm_cb_enter(void)
+{
+#if defined(CPU_SAML21)
```

Comment:
You can do

```suggestion
#if !defined(CPU_SAMD21)
```

I tried the patch on `same54-xpro` and the current consumption in STANDBY dropped from 2.66 mA to 0.22 mA while still being able to wake up from button press (`tests/periph_pm`)

Reply:
I made this on purpose ... I wanted to explicitly list all CPUs known to be compatible with this patch ;) So I think it is the best way - since you are able to test samd5x CPUs - to extend the list of supported CPUs instead of exclude others.

Do you like the concept at all? In that case I would rework this patch for `cpu/saml21` and `cpu/samd5x` and squash the commits.

Another question: If I add the `samd5x` HIBERNATE mode ... should I include this mode in the `gpio_pm_cb_enter()` function and also reduce the EIC clock?

Reply:
> I wanted to explicitly list all CPUs known to be compatible with this patch ;) So I think it is the best way - since you are able to test samd5x CPUs - to extend the list of supported CPUs instead of exclude others.

Yea but those lists of CPUs are a pain to hunt down when adding support for new family members.
I agree that `#if !defined(CPU_SAMD21)` is just as bad, but `#if defined(PM_SLEEPCFG_SLEEPMODE_STANDBY)` should do.
It's only defined by the more modern sam0 CPUs (everything newer than `samd2x`) and I would say if the same register with the same bits exists, it's safe to assume things are working pretty much the same. (that's how the other drivers in `sam0_common/periph` handle it too)

Since `pm.c` is not unified, it will need to be enabled for each family individually anyway.

>  If I add the `samd5x` HIBERNATE mode ... should I include this mode in the `gpio_pm_cb_enter()` function and also reduce the EIC clock?

I've tried `if (PM->SLEEPCFG.reg >= PM_SLEEPCFG_SLEEPMODE_STANDBY)` and it makes no difference in current consumption.
Looks like EIC is not powered in HIBERNATE or BACKUP either way, can't wake it by button press.

I also tried

```patch
--- a/cpu/sam0_common/periph/gpio.c
+++ b/cpu/sam0_common/periph/gpio.c
@@ -203,7 +203,7 @@ int gpio_init_int(gpio_t pin, gpio_mode_t mode, gpio_flank_t flank,
 #else /* CPU_FAM_SAML21 */
     /* enable clocks for the EIC module */
     MCLK->APBAMASK.reg |= MCLK_APBAMASK_EIC;
-    GCLK->PCHCTRL[EIC_GCLK_ID].reg = GCLK_PCHCTRL_CHEN | GCLK_PCHCTRL_GEN_GCLK0;
+    GCLK->PCHCTRL[EIC_GCLK_ID].reg = GCLK_PCHCTRL_CHEN | GCLK_PCHCTRL_GEN(SAM0_GCLK_32KHZ);
     /* disable the EIC module*/
     _EIC->CTRLA.reg = 0;
     EIC_SYNC();
```
and not doing `gpio_pm_cb_enter()` etc., with the same result (and 0.1mA higher STANDBY current)

Reply:
Thank you for your feedback!

Would you mind if I squash the current changes?

Reply:
sure, go ahead.